### PR TITLE
add dbginfo support for 64 bit entries (LDT,GDT) long mode

### PR DESCRIPTION
**Pull request by therealdreg**
- A mix from #32 and #32 
- Fixed some errors from #32 and #31
- Same code for: info LDT/GDT
- null entry display

Now it works as expected:

the correct thing

Tested on Windows x64 Host:
- Guest: Linux x86_64
- Guest: Windows x86 (32 bits)

@stlintel what do you think?

### CODE REVIEW

File: bochs/bx_debug/dbg_main.cc [hunk 1]
```diff
@@ -3130,67 +3146,76 @@ void bx_dbg_info_gdt_command(unsigned from, unsigned to)
...
+  dbg_printf("%s (base=0x" FMT_ADDRX ", limit=%d):\n", name, base, limit);
   for (unsigned n = from; n<=to; n++) {
-    Bit8u entry[8];
-    if (8*n + 7 > gdtr.limit) break;
-    if (bx_dbg_read_linear(dbg_cpu, gdtr.base + 8*n, 8, entry)) {
-      dbg_printf("GDT[0x%04x]=", n << 3);
+    Bit8u entry[16];
+    if (8*n + 7 > limit) break;
+    if (bx_dbg_read_linear(dbg_cpu, base + 8*n, 8, entry)) {
+      if (gdt)
```
Comment (stlintel) [inline bochs/bx_debug/dbg_main.cc:3154]: Please have the brackets placed same way as rest of the code

```
if (gdt) {
  dbg_printf("%s[0x%04x]=", name, n << 3);
}
else {
  dbg_printf("%s[0x%02x]=", name, n);
}

```

File: bochs/bx_debug/dbg_main.cc [hunk 2]
```diff
@@ -3130,67 +3146,76 @@ void bx_dbg_info_gdt_command(unsigned from, unsigned to)
...
+      else // LDT
+      {
+        dbg_printf("%s[0x%02x]=", name, n);
+      }
 
       Bit32u lo = (entry[3]  << 24) | (entry[2]  << 16) | (entry[1]  << 8) | (entry[0]);
       Bit32u hi = (entry[7]  << 24) | (entry[6]  << 16) | (entry[5]  << 8) | (entry[4]);
 
-      bx_dbg_print_descriptor(lo, hi);
+      if (0 == lo && 0 == hi)
```
Comment (stlintel) [inline bochs/bx_debug/dbg_main.cc:3166]: also here

File: bochs/bx_debug/dbg_main.cc [hunk 3]
```diff
@@ -3130,67 +3146,76 @@ void bx_dbg_info_gdt_command(unsigned from, unsigned to)
...
+            Bit32u lo2 = (entry[11] << 24) | (entry[10] << 16) | (entry[9] << 8) | (entry[8]);
+            Bit32u hi2 = (entry[15] << 24) | (entry[14] << 16) | (entry[13] << 8) | (entry[12]);
+
+            bx_dbg_print_descriptor64(lo, hi, lo2, hi2);
+          }
+          else {
+            dbg_printf("error: %sr+8*%d points to incomplete (size must be 16) descriptor, linear address 0x" FMT_ADDRX "\n", name, n, base);
+          }
+          continue;
+        }
```
Comment (stlintel) [inline bochs/bx_debug/dbg_main.cc:3188]: in my patch there was no else here
it was the else from if (longmode) and there was no code duplication

**Review by stlintel — CHANGES_REQUESTED** [review]

**Comment by stlintel** [comment]
Also I created a branch in Bochs repository for a reason.
If you would do the same I expect I was able to clone it and fix minor cosmetic issue shorting the ping-pong.
But I don't know how to clone from your fork :(

File: bochs/bx_debug/dbg_main.cc [hunk 4]
```diff
@@ -3110,18 +3110,34 @@ void bx_dbg_info_idt_command(unsigned from, unsigned to)
     dbg_printf("You can list individual entries with 'info idt [NUM]' or groups with 'info idt [NUM] [NUM]'\n");
 }
 
-void bx_dbg_info_gdt_command(unsigned from, unsigned to)
+void bx_dbg_info_lgdt_command(unsigned from, unsigned to, bool gdt)
 {
-  bx_dbg_global_sreg_t gdtr;
-  BX_CPU(dbg_cpu)->dbg_get_gdtr(&gdtr);
   bool all = 0;
+  char* name = "unk";
```
Comment (stlintel) [inline bochs/bx_debug/dbg_main.cc:3116]: const chat *name = gdt ? "gdt" : "ldt"

without any "unk" in the middle

**Review by stlintel — CHANGES_REQUESTED** [review]

File: bochs/bx_debug/dbg_main.cc (hunk 3 above)
Comment (therealdreg) [reply bochs/bx_debug/dbg_main.cc:3188]: this else its necessary (your code dont works as you expected,  it has some bugs)

**Comment by therealdreg** [comment]
> Also I created a branch in Bochs repository for a reason. If you would do the same I expect I was able to clone it and fix minor cosmetic issue shorting the ping-pong. But I don't know how to clone from your fork :(

I created my PR because your code have some bugs and a lot of code duplication with LDT logic etc.

File: bochs/bx_debug/dbg_main.cc (hunk 4 above)
Comment (therealdreg) [reply bochs/bx_debug/dbg_main.cc:3116]: ok, done

File: bochs/bx_debug/dbg_main.cc (hunk 1 above)
Comment (therealdreg) [reply bochs/bx_debug/dbg_main.cc:3154]: oh, crap, sorry, my fault (My fingers auto-Microsoft style haha) . Done!

File: bochs/bx_debug/dbg_main.cc (hunk 2 above)
Comment (therealdreg) [reply bochs/bx_debug/dbg_main.cc:3166]: done!

**Comment by therealdreg** [comment]
all your suggestions done, @stlintel what do you think now?

**Comment by therealdreg** [comment]
all your suggestions done.

New version compiled and tested in my local machine.

@stlintel what do you think now?

**Review by stlintel — APPROVED** [review]

**Comment by stlintel** [comment]
merged

**Pull request merged**